libcudacxx/test/libcudacxx/cuda/numeric/overflow.arithmetic/div_overflow.pass.cpp

```diff
@@ -0,0 +1,197 @@
+  [[maybe_unused]] constexpr auto lhs_max     = cuda::std::numeric_limits<Lhs>::max();
+  [[maybe_unused]] constexpr auto result_max  = cuda::std::numeric_limits<Result>::max();
+  [[maybe_unused]] constexpr auto result_min  = cuda::std::numeric_limits<Result>::min();
+  [[maybe_unused]] constexpr auto neg_lhs_min = cuda::uabs(lhs_min);
+  //--------------------------------------------------------------------------------------------------------------------
+  //  trivial cases
+  // 1. 0 / 1 -> should not overflow
+  test_div_overflow<Result>(Lhs{0}, Rhs{1}, false);
+
+  // 2. 1 / 1 -> should not overflow
+  test_div_overflow<Result>(Lhs{1}, Rhs{1}, false);
+
+  // nvcc 12.0 doesn't support these two special cases in a constexpr context
+#if _CCCL_CUDA_COMPILER(NVCC, !=, 12, 0)
+  // 3. 1 / -1 -> should overflow if the destination type is unsigned
+  // 4. -1 / 1 -> should overflow if the destination type is unsigned
+  test_nvcc_12_0_special_cases<Lhs, Rhs, Result>();
+#endif // _CCCL_CUDA_COMPILER(NVCC, !=, 12, 0)
+  if (!cuda::std::__cccl_default_is_constant_evaluated())
+  {
+    test_nvcc_12_0_special_cases<Lhs, Rhs, Result>();
+  }
+  // 5. 0 / -1
+  if constexpr (is_signed_v<Rhs>)
+  {
+    test_div_overflow<Result>(Lhs{0}, Rhs{-1}, false);
+  }
+  //--------------------------------------------------------------------------------------------------------------------
+  // min, max cases
+
+  // 6. max / 1 -> max >= result_max?
+  test_div_overflow<Result>(lhs_max, Rhs{1}, cuda::std::cmp_greater(lhs_max, result_max));
+
+  // nvcc 12.0 doesn't support this case in a constexpr context
+#if _CCCL_CUDA_COMPILER(NVCC, !=, 12, 0)
+  // 7. min / 1 -> min < result_min?
+  test_div_overflow<Result>(lhs_min, Rhs{1}, cuda::std::cmp_less(lhs_min, result_min));
+
+  if constexpr (is_signed_v<Lhs> && is_signed_v<Rhs>)
+  {
+    // 8. min / -1
+    // when the result type is larger than the common type, min / -1 produces a valid result
+    bool overflow = cuda::std::cmp_greater(neg_lhs_min, result_max);
+    test_div_overflow<Result>(lhs_min, Rhs{-1}, overflow, !overflow, static_cast<Result>(neg_lhs_min));
+
+    // 9. min / -2
+    bool overflow2 = cuda::std::cmp_greater(neg_lhs_min / 2, result_max);
+    test_div_overflow<Result>(lhs_min, Rhs{-2}, overflow2, !overflow2, static_cast<Result>(neg_lhs_min / 2));
+  }
+#endif // _CCCL_CUDA_COMPILER(NVCC, !=, 12, 0)
+}
+
+template <typename Lhs, typename Rhs>
+__host__ __device__ constexpr void test_type()
+{
+  test_type<Lhs, Rhs, cuda::std::common_type_t<Lhs, Rhs>>();
+  test_type<Lhs, Rhs, unsigned>();
+  test_type<Lhs, Rhs, int>();
+  // instantiation of all tests is very expensive. clang hits "constexpr evaluation hit maximum step limit"
+#if _CCCL_COMPILER(GCC)
+  test_type<Lhs, Rhs, long long>();
+  test_type<Lhs, Rhs, unsigned long long>();
+  test_type<Lhs, Rhs, signed char>();
+  test_type<Lhs, Rhs, unsigned char>();
+#endif
```

Comment:
Nitpick: We do add the corresponding comments at the end of conditional compilation blocks